ste/sourceInfoProvider-GCP.go

```diff
@@ -0,0 +1,167 @@
+		return nil, err
+	}
+	p.gcpURLParts, err = common.NewGCPURLParts(*p.rawSourceURL)
+	if err != nil {
+		return nil, err
+	}
+
+	p.gcpClient, err = gcpClientFactory.GetGCPClient(
+		p.jptm.Context(),
+		common.CredentialInfo{
+			CredentialType:    common.ECredentialType.GoogleAppCredentials(),
+			GCPCredentialInfo: common.GCPCredentialInfo{},
+		},
+		common.CredentialOpOptions{
+			LogInfo:  func(str string) { p.jptm.Log(pipeline.LogInfo, str) },
+			LogError: func(str string) { p.jptm.Log(pipeline.LogError, str) },
+			Panic:    func(err error) { panic(err) },
+		})
+	if err != nil {
+		return nil, err
+	}
+	glcm := common.GetLifecycleMgr()
+	jsonKey, err = ioutil.ReadFile(glcm.GetEnvironmentVariable(common.EEnvironmentVariable.GoogleAppCredentials()))
+	if err != nil {
+		return nil, fmt.Errorf("Cannot read JSON key file. Please verify you have correctly set GOOGLE_APPLICATION_CREDENTIALS environment variable")
+	}
+	return &p, nil
+}
+
+func (p *gcpSourceInfoProvider) PreSignedSourceURL() (*url.URL, error) {
+
+	conf, err := google.JWTConfigFromJSON(jsonKey)
+	if err != nil {
+		return nil, fmt.Errorf("Could not get config from json key. Error: %v", err)
+	}
+	opts := &storage.SignedURLOptions{
+		Scheme:         storage.SigningSchemeV4,
+		Method:         "GET",
+		GoogleAccessID: conf.Email,
+		PrivateKey:     conf.PrivateKey,
+		Expires:        time.Now().Add(defaultPresignExpires),
+	}
+	u, err := storage.SignedURL(p.gcpURLParts.BucketName, p.gcpURLParts.ObjectKey, opts)
+
+	if err != nil {
+		return nil, fmt.Errorf("Unable to Generate Signed URL for given GCP Object: %v", err)
+	}
+
+	parsedURL, err := url.Parse(u)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to parse signed URL: %v", err)
+	}
+
+	return parsedURL, nil
+}
+
+func (p *gcpSourceInfoProvider) Properties() (*SrcProperties, error) {
+	srcProperties := SrcProperties{
+		SrcHTTPHeaders: p.transferInfo.SrcHTTPHeaders,
+		SrcMetadata:    p.transferInfo.SrcMetadata,
+	}
+	if p.transferInfo.S2SGetPropertiesInBackend {
+		objectInfo, err := p.gcpClient.Bucket(p.gcpURLParts.BucketName).Object(p.gcpURLParts.ObjectKey).Attrs(p.jptm.Context())
+		if err != nil {
+			return nil, err
+		}
+		oie := common.GCPObjectInfoExtension{ObjectInfo: *objectInfo}
+		srcProperties = SrcProperties{
+			SrcHTTPHeaders: common.ResourceHTTPHeaders{
+				ContentType:        objectInfo.ContentType,
+				ContentEncoding:    oie.ContentEncoding(),
+				ContentDisposition: oie.ContentDisposition(),
+				ContentLanguage:    oie.ContentLanguage(),
+				CacheControl:       oie.CacheControl(),
+				ContentMD5:         oie.ContentMD5(),
+			},
+			SrcMetadata: oie.NewCommonMetadata(),
+		}
+	}
+	resolvedMetadata, err := p.handleInvalidMetadataKeys(srcProperties.SrcMetadata)
+
+	if err != nil {
+		return nil, err
+	}
+	srcProperties.SrcMetadata = resolvedMetadata
+	return &srcProperties, nil
+}
+
+func (p *gcpSourceInfoProvider) handleInvalidMetadataKeys(m common.Metadata) (common.Metadata, error) {
```

Comment:
This code looks identical to the S3 source info provider, consider consolidating the two methods if possible.

Reply:
Probably gonna leave this. This could be made a method of common.Metadata, but that is logically wrong. Suggestions?